pkg/iac/scanners/terraform/ignore_test.go

```diff
@@ -1,395 +1,280 @@
+			source: `
 //tfsec:ignore:*
 //tfsec:ignore:a
 //tfsec:ignore:b
 //tfsec:ignore:c
 //tfsec:ignore:d
-resource "bad" "my-rule" {
-   secure = false 
-}
+resource "aws_s3_bucket" "test" {}
 `,
 			assertLength: 0,
 		},
 		{
-			name: "IgnoreLineAboveTheLine",
-			inputOptions: `
-resource "bad" "my-rule" {
-	# tfsec:ignore:aws-service-abc123
-    secure = false
+			name: "rule above the finding",
+			source: `
+resource "aws_s3_bucket" "test" {
+	# tfsec:ignore:aws-s3-non-empty-bucket
+    bucket = ""
 }
 `,
 			assertLength: 0,
 		},
 		{
-			name: "IgnoreWithExpDateIfDateBreachedThenDontIgnore",
-			inputOptions: `
-resource "bad" "my-rule" {
-    secure = false # tfsec:ignore:aws-service-abc123:exp:2000-01-02
+			name: "rule with breached expiration date",
+			source: `
+resource "aws_s3_bucket" "test" {
+    bucket = "" # tfsec:ignore:aws-s3-non-empty-bucket:exp:2000-01-02
 }
 `,
 			assertLength: 1,
 		},
 		{
-			name: "IgnoreWithExpDateIfDateNotBreachedThenIgnoreIgnore",
-			inputOptions: `
-resource "bad" "my-rule" {
-    secure = false # tfsec:ignore:aws-service-abc123:exp:2221-01-02
+			name: "rule with unbreached expiration date",
+			source: `
+resource "aws_s3_bucket" "test" {
+    bucket = "" # tfsec:ignore:aws-s3-non-empty-bucket:exp:2221-01-02
 }
 `,
 			assertLength: 0,
 		},
 		{
-			name: "IgnoreWithExpDateIfDateInvalidThenDropTheIgnore",
-			inputOptions: `
-resource "bad" "my-rule" {
-   secure = false # tfsec:ignore:aws-service-abc123:exp:2221-13-02
+			name: "rule with invalid expiration date",
+			source: `
+resource "aws_s3_bucket" "test" {
+   bucket = "" # tfsec:ignore:aws-s3-non-empty-bucket:exp:2221-13-02
 }
 `,
 			assertLength: 1,
 		},
 		{
-			name: "IgnoreAboveResourceBlockWithExpDateIfDateNotBreachedThenIgnoreIgnore",
-			inputOptions: `
-#tfsec:ignore:aws-service-abc123:exp:2221-01-02
-resource "bad" "my-rule" {
-}
-`,
-			assertLength: 0,
-		},
-		{
-			name: "IgnoreAboveResourceBlockWithExpDateAndMultipleIgnoresIfDateNotBreachedThenIgnoreIgnore",
-			inputOptions: `
-# tfsec:ignore:aws-service-abc123:exp:2221-01-02
-resource "bad" "my-rule" {
-	
-}
-`,
-			assertLength: 0,
-		},
-		{
-			name: "IgnoreForImpliedIAMResource",
-			inputOptions: `
-terraform {
-  required_version = "~> 1.1.6"
-
-  required_providers {
-    aws = {
-      source  = "hashicorp/aws"
-      version = "~> 3.48"
-    }
-  }
-}
-
-# Retrieve an IAM group defined outside of this Terraform config.
-
-# tfsec:ignore:aws-iam-enforce-mfa
-data "aws_iam_group" "externally_defined_group" {
-  group_name = "group-name" # tfsec:ignore:aws-iam-enforce-mfa
-}
-
-# Create an IAM policy and attach it to the group.
-
-# tfsec:ignore:aws-iam-enforce-mfa
-resource "aws_iam_policy" "test_policy" {
-  name   = "test-policy"                                 # tfsec:ignore:aws-iam-enforce-mfa
-  policy = data.aws_iam_policy_document.test_policy.json # tfsec:ignore:aws-iam-enforce-mfa
```

Comment:
The `AVD-AWS-0123` (`aws-iam-enforce-mfa`) check will never be triggered for this input even without ignore rules.

Reply:
Why is that? If so is the case, was the assertion incorrect?

Reply:
`aws-iam-enforce-mfa` checks that the condition key is `aws:MultiFactorAuthPresent`. but the test data does not contain this. I also ran trivy on the test case, but it didn't produce any results:
```bash
trivy conf main.tf
2024-11-11T15:36:25+06:00       INFO    [misconfig] Misconfiguration scanning is enabled
2024-11-11T15:36:26+06:00       INFO    [terraform scanner] Scanning root module        file_path="."
2024-11-11T15:36:26+06:00       INFO    Detected config files   num=1
```

Reply:
Can you share your main.tf? If it doesn't detect it, it's a bug and should be looked into.

Reply:
```bash
❯ cat main.tf
terraform {
  required_version = "~> 1.1.6"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.48"
    }
  }
}

# Retrieve an IAM group defined outside of this Terraform config.


data "aws_iam_group" "externally_defined_group" {
  group_name = "group-name" 
}

# Create an IAM policy and attach it to the group.


resource "aws_iam_policy" "test_policy" {
  name   = "test-policy"                                 
  policy = data.aws_iam_policy_document.test_policy.json 
}


resource "aws_iam_group_policy_attachment" "test_policy_attachment" {
  group      = data.aws_iam_group.externally_defined_group.group_name 
  policy_arn = aws_iam_policy.test_policy.arn                         
}


data "aws_iam_policy_document" "test_policy" {
  statement {
    sid = "PublishToCloudWatch" 
    actions = [
      "cloudwatch:PutMetricData", 
    ]
    resources = ["*"] 
  }
}%
```

Reply:
Why is this a bug? The check should not work for this test data, because it does not match the check.

Reply:
My inattention, the check didn't trigger because the group was declared as a data block, not a resource. Added the corrected test case back in.